torch/_functorch/_activation_offloading/activation_offloading.py

```diff
@@ -0,0 +1,497 @@
+    for node in fwd_module.graph.nodes:
+        if node.meta.get("should_offload", False) and can_offload(
+            node, fwd_outputs, model_outputs, static_lifetime_input_nodes
+        ):
+            node.meta["saved_for_offloading"] = True
+            node.meta["original_device"] = node.meta["val"].device
+            should_perform_offloading = True
+
+    return should_perform_offloading
+
+
+def offload_chosen_sets(
+    fwd_module: fx.GraphModule,
+    bwd_module: fx.GraphModule,
+) -> None:
+    """
+    Add offload and reload nodes to the forward and backward graphs.
+    This function adds device_put operations without any stream handling.
+
+    Args:
+        fwd_module: Forward module graph
+        bwd_module: Backward module graph
+    """
+
+    # Add offload nodes in forward graph
+    offload_activation_fw(fwd_module.graph)
+
+    # Update backward graph inputs to be offloaded tensors
+    bwd_inputs: dict[str, fx.Node] = {
+        node.name: node for node in bwd_module.graph.find_nodes(op="placeholder")
+    }
+    for fwd_node in fwd_module.graph.find_nodes(op="output")[0].args[0]:
+        if CPU_OFFLOAD_PREFIX not in fwd_node.name:
+            continue
+
+        bwd_node: fx.Node = bwd_inputs[fwd_node.name.replace(CPU_OFFLOAD_PREFIX, "")]
+        with bwd_module.graph.inserting_after(bwd_node):
+            bwd_offload_node: fx.Node = bwd_module.graph.placeholder(name=fwd_node.name)
+
+        bwd_offload_node.meta.update(fwd_node.meta)
+        bwd_offload_node.meta["saved_for_offloading"] = True
+        bwd_offload_node.meta["original_device"] = bwd_node.meta["val"].device
+        bwd_node.replace_all_uses_with(bwd_offload_node)
+        bwd_module.graph.erase_node(bwd_node)
+
+    # Add reload nodes in backward graph
+    reload_activation_bw(bwd_module.graph)
+
+
+def add_forward_offload_stream_ops(graph: fx.Graph) -> None:
+    """
+    Add stream operations for forward pass CPU offloading.
+
+    Pattern: record_event → fork → wait_event → device_put → record_event_2 → join → wait_event_2
+
+    This ensures that:
+    1. Offloading waits for the last use to complete (record_event on default stream)
+    2. Offloading happens on a separate stream (fork → wait_event → device_put)
+    3. Execution returns to the default stream after offloading and
+       waits for offload to complete (record_event_2 → join → wait_event_2)
+
+    NOTE: For stream optimization and overlapping compute with communication,
+          the "wait_event_2" ops can be sinked to the end of the graph.
+
+    Args:
+        graph: The forward graph to modify
+    """
+
+    # Find all CPU offload nodes
+    offload_nodes: list[fx.Node] = [
+        node
+        for node in graph.nodes
+        if CPU_OFFLOAD_PREFIX in node.name and node.op == "call_function"
+    ]
+    if not offload_nodes:
+        return
+
+    # Get default stream id and offload stream id
+    default_stream_id: int = get_current_stream(
```

Comment:
nit: Maybe rename mentions of "default" stream to ambient/current stream (here and other places)?